### PR TITLE
added application-test.yml for H2, added dependency to pom.xml, added…

### PR DESCRIPTION
**Pull request by MarvelGa**
added application-test.yml for H2, added dependency to pom.xml, added RoleServiceImplIntegrationTest

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -45,6 +45,12 @@
             <artifactId>spring-boot-starter-test</artifactId>
             <scope>test</scope>
         </dependency>
+
+        <dependency>
+            <groupId>com.h2database</groupId>
+            <artifactId>h2</artifactId>
+            <scope>runtime</scope>
```
Comment (dmitriychernov7) [inline pom.xml]: it doesn't need to be runtime. it should be "test"

File: src/test/java/com/exadel/discountwebapp/integrationtest/service/RoleServiceImplIntegrationTest.java [hunk 2]
```diff
@@ -0,0 +1,79 @@
+package com.exadel.discountwebapp.integrationtest.service;
```
Comment (ahlukhouski2) [inline src/test/java/com/exadel/discountwebapp/integrationtest/service/RoleServiceImplIntegrationTest.java]: com.exadel.discountwebapp.role ?

File: src/test/java/com/exadel/discountwebapp/integrationtest/service/RoleServiceImplIntegrationTest.java [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+import javax.transaction.Transactional;
+import java.time.LocalDateTime;
+import java.util.List;
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+
+@Transactional
+@SpringBootTest(classes = DiscountWebApplication.class)
+@AutoConfigureTestDatabase(connection = EmbeddedDatabaseConnection.H2)
+public class RoleServiceImplIntegrationTest {
```
Comment (ahlukhouski2) [inline src/test/java/com/exadel/discountwebapp/integrationtest/service/RoleServiceImplIntegrationTest.java]: RoleServiceIntegrationTest

File: src/test/java/com/exadel/discountwebapp/integrationtest/service/RoleServiceImplIntegrationTest.java [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+
```
Comment (ahlukhouski2) [inline src/test/java/com/exadel/discountwebapp/integrationtest/service/RoleServiceImplIntegrationTest.java]: @Autowired

File: src/test/java/com/exadel/discountwebapp/integrationtest/service/RoleServiceImplIntegrationTest.java [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+import org.springframework.test.annotation.Rollback;
+
+import javax.transaction.Transactional;
+import java.time.LocalDateTime;
+import java.util.List;
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+
+@Transactional
+@SpringBootTest(classes = DiscountWebApplication.class)
```
Comment (ahlukhouski2) [inline src/test/java/com/exadel/discountwebapp/integrationtest/service/RoleServiceImplIntegrationTest.java]: Try remove classes = DiscountWebApplication.class, should work without it

File: src/test/java/com/exadel/discountwebapp/integrationtest/service/RoleServiceImplIntegrationTest.java [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+@AutoConfigureTestDatabase(connection = EmbeddedDatabaseConnection.H2)
```
Comment (ahlukhouski2) [inline src/test/java/com/exadel/discountwebapp/integrationtest/service/RoleServiceImplIntegrationTest.java]: Try remove @AutoConfigureTestDatabase(connection = EmbeddedDatabaseConnection.H2)

File: src/test/java/com/exadel/discountwebapp/integrationtest/service/RoleServiceImplIntegrationTest.java [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+import org.springframework.boot.test.context.SpringBootTest;
+import org.springframework.test.annotation.Rollback;
+
+import javax.transaction.Transactional;
+import java.time.LocalDateTime;
+import java.util.List;
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+
+@Transactional
```
Comment (ahlukhouski2) [inline src/test/java/com/exadel/discountwebapp/integrationtest/service/RoleServiceImplIntegrationTest.java]: We can clean up database after every test with SQL cleanup script (@Sql( ... cleanup.sql ))

File: src/test/java/com/exadel/discountwebapp/integrationtest/service/RoleServiceImplIntegrationTest.java [hunk 8]
```diff
@@ -0,0 +1,79 @@
...
+
+    private RoleRepository roleRepository;
+
+    private RoleService roleService;
+
+    private Role roleUser;
+    private Role roleAdmin;
+
+    @Autowired
+    public RoleServiceImplIntegrationTest(RoleMapper roleMapper, RoleRepository roleRepository, RoleService roleService) {
```
Comment (ahlukhouski2) [inline src/test/java/com/exadel/discountwebapp/integrationtest/service/RoleServiceImplIntegrationTest.java]: can be removed

File: src/test/java/com/exadel/discountwebapp/integrationtest/service/RoleServiceImplIntegrationTest.java [hunk 9]
```diff
@@ -0,0 +1,79 @@
...
+                .build();
+        roleUser = roleRepository.save(roleUser);
+
+        roleAdmin = Role.builder()
+                .id(2L)
+                .created(LocalDateTime.now())
+                .modified(LocalDateTime.now())
+                .name("ADMIN")
+                .build();
+        roleAdmin = roleRepository.save(roleAdmin);
```
Comment (ahlukhouski2) [inline src/test/java/com/exadel/discountwebapp/integrationtest/service/RoleServiceImplIntegrationTest.java]: TransactionTemplate can be used to run inside transaction

Comment (ahlukhouski2) [inline src/test/java/com/exadel/discountwebapp/integrationtest/service/RoleServiceImplIntegrationTest.java]: saveAll() ?

File: src/test/java/com/exadel/discountwebapp/integrationtest/service/RoleServiceImplIntegrationTest.java [hunk 10]
```diff
@@ -0,0 +1,79 @@
...
+        this.roleMapper = roleMapper;
+        this.roleRepository = roleRepository;
+        this.roleService = roleService;
+    }
+
+    @BeforeEach
+    void setUp() {
+        roleUser = Role.builder()
+                .id(1L)
+                .created(LocalDateTime.now())
```
Comment (ahlukhouski2) [inline src/test/java/com/exadel/discountwebapp/integrationtest/service/RoleServiceImplIntegrationTest.java]: created/modified is redundant here ?

File: src/test/java/com/exadel/discountwebapp/integrationtest/service/RoleServiceImplIntegrationTest.java [hunk 11]
```diff
@@ -0,0 +1,79 @@
...
+        roleAdmin = Role.builder()
+                .id(2L)
+                .created(LocalDateTime.now())
+                .modified(LocalDateTime.now())
+                .name("ADMIN")
+                .build();
+        roleAdmin = roleRepository.save(roleAdmin);
+    }
+
+    @Rollback
```
Comment (ahlukhouski2) [inline src/test/java/com/exadel/discountwebapp/integrationtest/service/RoleServiceImplIntegrationTest.java]: can be removed

File: src/test/java/com/exadel/discountwebapp/integrationtest/service/RoleServiceImplIntegrationTest.java [hunk 12]
```diff
@@ -0,0 +1,79 @@
...
+        List<RoleResponseVO> actual = roleService.getAllRoles();
+        assertEquals(expected, actual);
+    }
+
+    @Rollback
+    @Test
+    void shouldGetRoleById() {
+        var expected = roleMapper.toVO(roleAdmin);
+        var id = roleAdmin.getId();
+        var actual = roleService.getRoleById(2);
```
Comment (ahlukhouski2) [inline src/test/java/com/exadel/discountwebapp/integrationtest/service/RoleServiceImplIntegrationTest.java]: id?

File: src/test/resources/application-test.yml [hunk 13]
```diff
@@ -0,0 +1,7 @@
+spring:
+  datasource:
+    driver-class-name: org.h2.Driver
+    url: jdbc:h2:mem:test;DB_CLOSE_DELAY=-1;DATABASE_TO_UPPER=false;
```
Comment (ahlukhouski2) [inline src/test/resources/application-test.yml]: test?
DATABASE_TO_UPPER=false; ?

File: src/test/java/com/exadel/discountwebapp/role/RoleServiceIntegrationTest.java [hunk 14]
```diff
@@ -0,0 +1,67 @@
...
+    @Autowired
+    private TransactionTemplate transactiontemplate;
+
+    private Role roleUser;
+    private Role roleAdmin;
+
+    @BeforeEach
+    void setUp() {
+        transactiontemplate.execute(status -> {
+            roleUser = Role.builder()
```
Comment (ahlukhouski2) [inline src/test/java/com/exadel/discountwebapp/role/RoleServiceIntegrationTest.java]: creating roles can be done outside the transaction scope
a transaction should be kept as short as possible

File: src/test/java/com/exadel/discountwebapp/role/RoleServiceIntegrationTest.java [hunk 15]
```diff
@@ -0,0 +1,67 @@
...
+                    .name("USER")
+                    .build();
+
+            roleAdmin = Role.builder()
+                    .id(2L)
+                    .name("ADMIN")
+                    .build();
+
+            roleRepository.saveAll(List.of(roleUser, roleAdmin));
+            return "";
```
Comment (ahlukhouski2) [inline src/test/java/com/exadel/discountwebapp/role/RoleServiceIntegrationTest.java]: just write 
return roleRepository.saveAll(List.of(roleUser, roleAdmin));

File: src/test/java/com/exadel/discountwebapp/role/RoleServiceIntegrationTest.java [hunk 16]
```diff
@@ -0,0 +1,67 @@
...
+import org.junit.jupiter.api.Test;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.boot.test.context.SpringBootTest;
+import org.springframework.test.context.jdbc.Sql;
+import org.springframework.transaction.support.TransactionTemplate;
+import java.util.List;
+import static org.junit.jupiter.api.Assertions.assertEquals;
+
+@Sql(executionPhase = Sql.ExecutionPhase.AFTER_TEST_METHOD, scripts =
+        "classpath:cleanup.sql")
```
Comment (ahlukhouski2) [inline src/test/java/com/exadel/discountwebapp/role/RoleServiceIntegrationTest.java]: let's write it in one line

File: src/test/resources/cleanup.sql [hunk 17]
```diff
@@ -0,0 +1,4 @@
+USE discount_web_app;
```
Comment (ahlukhouski2) [inline src/test/resources/cleanup.sql]: let's move this file to a separate directory
resources/sql/cleanup.sql for instance

File: src/test/java/com/exadel/discountwebapp/role/RoleServiceIntegrationTest.java [hunk 18]
```diff
@@ -0,0 +1,67 @@
...
+public class RoleServiceIntegrationTest {
+
+    @Autowired
+    private RoleMapper roleMapper;
+    @Autowired
+    private RoleRepository roleRepository;
+    @Autowired
+    private RoleService roleService;
+    @Autowired
+    private TransactionTemplate transactiontemplate;
```
Comment (ahlukhouski2) [inline src/test/java/com/exadel/discountwebapp/role/RoleServiceIntegrationTest.java]: transactionTemplate

File: src/test/java/com/exadel/discountwebapp/role/RoleServiceIntegrationTest.java [hunk 19]
```diff
@@ -0,0 +1,67 @@
...
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.boot.test.context.SpringBootTest;
+import org.springframework.test.context.jdbc.Sql;
+import org.springframework.transaction.support.TransactionTemplate;
+import java.util.List;
+import static org.junit.jupiter.api.Assertions.assertEquals;
+
+@Sql(executionPhase = Sql.ExecutionPhase.AFTER_TEST_METHOD, scripts =
+        "classpath:cleanup.sql")
+@SpringBootTest()
```
Comment (ahlukhouski2) [inline src/test/java/com/exadel/discountwebapp/role/RoleServiceIntegrationTest.java]: () can be removed

**Review by ahlukhouski2 — APPROVED** [review]

File: src/test/java/com/exadel/discountwebapp/role/RoleServiceIntegrationTest.java [hunk 20]
```diff
@@ -0,0 +1,66 @@
...
+        List<RoleResponseVO> actual = roleService.getAllRoles();
+        assertEquals(expected, actual);
+    }
+
+    @Test
+    void shouldGetRoleById() {
+        var expected = roleMapper.toVO(roleAdmin);
+        var id = roleAdmin.getId();
+        var actual = roleService.getRoleById(id);
+        assertEquals(expected, actual);
```
Comment (ahlukhouski2) [inline src/test/java/com/exadel/discountwebapp/role/RoleServiceIntegrationTest.java]: assertEquals(roleAdmin.getName(), actual.getName());
assertEquals(roleAdmin.getId(), actual.getId());

**Pull request merged**